### PR TITLE
LPS-145927

### PR DESCRIPTION
**Pull request by marianoalvarosaiz**

### CODE REVIEW

File: portal-impl/src/com/liferay/portal/language/LanguageResources.java [hunk 1]
```diff
@@ -303,7 +303,7 @@ private static Locale _getSuperLocale(Locale locale) {
 			languageMap = Collections.emptyMap();
 		}
 
-		_languageMaps.put(locale, languageMap);
+		_languageMaps.putIfAbsent(locale, languageMap);
```
Comment (tinatian) [inline portal-impl/src/com/liferay/portal/language/LanguageResources.java:306]: @marianoalvarosaiz How is this change related to the fix?

Tina

Comment (tinatian) [reply portal-impl/src/com/liferay/portal/language/LanguageResources.java:306]: @marianoalvarosaiz This code is not going to related to race condition. If you take a deep at how this is used, you would see that the language map put in here is always an empty map, and no one is going to modify the map later, all the usages is to read the map. I am going to skip this commit.

Tina.

File: portal-kernel/test/unit/com/liferay/portal/kernel/util/LocaleUtilTest.java [hunk 2]
```diff
@@ -119,4 +119,19 @@ public void testFromLanguageIdConsistency() {
 		Assert.assertNull(LocaleUtil.fromLanguageId("de_DE", true, false));
 	}
 
+	@Test
+	public void testFromLanguageIdLocaleIsCreatedAndRetrievableWhenNoValidationDone() {
```
Comment (tinatian) [inline portal-kernel/test/unit/com/liferay/portal/kernel/util/LocaleUtilTest.java:140]: @marianoalvarosaiz The test name is about "No Validation", but the parameter passed into LanguageUtil is true. Also, please use Assert.assertSame here instead of ==

File: portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java [hunk 3]
```diff
@@ -287,16 +287,29 @@ private Locale _fromLanguageId(
...
-		if (validate && !LanguageUtil.isAvailableLocale(locale)) {
-			if (_log.isWarnEnabled()) {
-				_log.warn(languageId + " is not a valid language id");
-			}
+		if (validate) {
+			boolean languageCode = false;
 
-			if (useDefault) {
-				return _locale;
+			if ((languageId.indexOf(CharPool.UNDERLINE) < 0) &&
```
Comment (tinatian) [inline portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java:293]: @marianoalvarosaiz There was a part of logic to adjust languageId, and your logic here did not include it. 
`if (languageId.equals("zh-Hans-CN")) {
			languageId = "zh_CN";
		}
		else if (languageId.equals("zh-Hant-TW")) {
			languageId = "zh_TW";
		}
		else {
			languageId = StringUtil.replace(
				languageId, CharPool.MINUS, CharPool.UNDERLINE);
		}`

**Comment by tinatian** [comment]
@marianoalvarosaiz Please see my messages above and resend.

Thanks.
Tina.

**Comment by marianoalvarosaiz** [comment]
see https://github.com/liferay-core-infra/liferay-portal/pull/342

File: portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java (hunk 3 above)
Comment (marianoalvarosaiz) [reply portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java:293]: @tinatian already covered by com.liferay.portal.kernel.util.LocaleUtilTest.testFromLanguageIdBCP47()

Comment (marianoalvarosaiz) [reply portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java:293]: @tinatian it is included when I first call `Locale locale = _fromLanguageId(languageId);` After that I only want to check if languageId is a language code (everything that doesn't include a **-** or **_**) which _zh-Hans-CN_ is not so it is cover by the first method call.
